### DISCUSSION TITLE
Asynchronous Dispatch in JAX

### DISCUSSION THREAD

**Discussion by goktugguvercin** [jax-ml/jax#9895] [Q&A]
Hello JAX Community;

While testing asynchronous dispatch in JAX, I noticed that total duration for dot product of two matrices with and without "block_until_ready()" is almost same. It is like asynchronous dispatch does not work. I cannot see remarkable difference between two measured time slots. I executed the following code on Kaggle platform, and I started to ask myself whether asynchronous dispatch is usable for only GPUs, because I have only CPU.

```python
x = random.uniform(random.PRNGKey(0), (1000, 1000))
%time jnp.dot(x, x)
%time jnp.dot(x, x).block_until_ready()
```

**Time Results:**

1. ** total: 59.2 ms **
2. ** total: 64.3 ms **

**Comment by jakevdp** [comment]
I think the asynchronous dispatch may only be important on GPU / TPU; for example, this is the output I get for your code on a standard Colab GPU runtime:
```
CPU times: user 101 µs, sys: 48 µs, total: 149 µs
Wall time: 161 µs
CPU times: user 1.03 ms, sys: 3 µs, total: 1.03 ms
Wall time: 1.03 ms
```

**Reply by mattjj** [reply]
Async dispatch is also kind of complex; dispatch is not _guaranteed_ to be async, but it may be async. For example, on CPU small computations (by FLOP count, as a heuristic) are executed on the main thread (rather than asynchronously on another thread); see [1](https://github.com/tensorflow/tensorflow/blob/bcb223e93f1365759012a9d2da8090341ead96f8/tensorflow/compiler/xla/pjrt/tfrt_cpu_pjrt_client.cc#L1316) and [2](https://github.com/tensorflow/tensorflow/blob/bcb223e93f1365759012a9d2da8090341ead96f8/tensorflow/compiler/xla/pjrt/tfrt_cpu_pjrt_client.cc#L1608). Also there's backpressure; only a certain (backend-dependent) number of in-flight async computations are allowed at once, so when doing timing loops like `%timeit` that queue can fill up and one is effectively blocking the main thread until computations finish (i.e. in steady-state it's like calling `block_until_ready`).

@zhangqiaorjc any detail to add here?

**Reply by zhangqiaorjc** [reply]
Yes, essentially users should assume async dispatch and use `block_until_ready` for benchmarks. But on CPU backend specifically, we found that for cheap computations, async hurts performance so we may choose to dispatch/compute those in the same thread.

We determine if a CPU computation using HLO cost estimate

```c++
  // The magic constant 1000 is determined by correlating computation with flop
  // estimate. It is a crude heuristic to find computation less than the thread
  // context switch time (~5us).
  cheap_computation_ = hlo_cost_analysis->flop_count() < 1000;
```
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/pjrt/tfrt_cpu_pjrt_client.cc#L1312-L1316

And whether dispatch is to the same thread or to a threadpool is done here
https://github.com/tensorflow/tensorflow/blob/master/tensorflow/compiler/xla/pjrt/tfrt_cpu_pjrt_client.cc#L1608-L1686

**Reply by zhangqiaorjc** [reply]
Twist...we have a bug in estimating flop counts, i missed a line and we have been getting `0` for flop counts. Will fix now, and it will be available in the next jaxlib release.

```c++
auto hlo_cost_analysis =
     std::make_unique<HloCostAnalysis>(cpu::CpuExecutable::ShapeSizeBytes);
+  TF_RETURN_IF_ERROR(cpu_executable->module().entry_computation()->Accept(
+      hlo_cost_analysis.get()));
```

**Reply by zhangqiaorjc** [reply]
Here's a simple way to find out the flop estimates from XLA for any JAX computation

```python
x = np.random.uniform(size=(1000, 1000)).astype(np.float32)

def f(x):
  return jnp.dot(x, x)

m = jax.xla_computation(f)(x).as_hlo_module()
client = jax.lib.xla_bridge.get_backend()
analysis = jax.lib.xla_client._xla.hlo_module_cost_analysis(client, m)
flops = analysis['flops']
print(flops)
```
Notably, the flops value here is `2000000000.0`.

**Reply by YouJiacheng** [reply]
> Twist...we have a bug in estimating flop counts, i missed a line and we have been getting `0` for flop counts. Will fix now, and it will be available in the next jaxlib release.
> 
> ```c++
> auto hlo_cost_analysis =
>      std::make_unique<HloCostAnalysis>(cpu::CpuExecutable::ShapeSizeBytes);
> +  TF_RETURN_IF_ERROR(cpu_executable->module().entry_computation()->Accept(
> +      hlo_cost_analysis.get()));
> ```

Does the bug only affect CPU backend?

**Reply by goktugguvercin** [reply]
I guess I have no question in my mind, thank you @mattjj  and @zhangqiaorjc so much for your elegant explanation.

**Reply by zhangqiaorjc** [reply]
@YouJiacheng yes, only for CPU backend; for GPU and TPU, we always dispatch async from host to device with a configurable max inflight computations (32 is the default). CPU is special because sync dispatch for cheap computations is a performance optimization.
